test/EndToEnd/tests/PackageNameSpaceTests.ps1

```diff
@@ -0,0 +1,171 @@
+        '<packages>
+            <package id="SolutionLevelPkg" version="1.0.0" targetFramework="net461" />
+    </packages>' | out-file $packagesConfigPath
+    
+        # Act
+        Build-Solution
+    
+        # Assert
+        $packagesFolder = Join-Path $solutionDirectory "packages"
+        $solutionLevelPkgNupkgFolder = Join-Path $packagesFolder "SolutionLevelPkg.1.0.0"
+        Assert-PathExists(Join-Path $solutionLevelPkgNupkgFolder "SolutionLevelPkg.1.0.0.nupkg")
+        
+        $errorlist = Get-Errors
+        Assert-AreEqual 0 $errorlist.Count
+    }
+    finally {
+        Remove-Item $nugetConfigPath
+    }
+}
+
+function Test-PackageNamespaceRestore-WithMultipleFeedsWithIdenticalPackages-RestoresCorrectPackage
+{
+    param($context)
+
+    # Arrange
+    $repoDirectory = Join-Path $OutputPath "CustomPackages"
+    $opensourceRepo = Join-Path $repoDirectory "opensourceRepo"
+    $privateRepo = Join-Path $repoDirectory "privateRepo"
+    $nugetConfigPath = Join-Path $OutputPath 'nuget.config'
+
+	$settingFileContent =@"
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+    <packageSources>
+    <clear />
+    <add key="OpensourceRepository" value="{0}" />
+    <add key="PrivateRepository" value="{1}" />
+    </packageSources>
+    <packageNamespaces>
+        <packageSource key="PrivateRepository">
+            <namespace id="Contoso.MVC.*" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>
+"@
+    try {
+        # We have to create config file before creating solution, otherwise it's not effective for new solutions.
+        $settingFileContent -f $opensourceRepo,$privateRepo | Out-File -Encoding "UTF8" $nugetConfigPath
+
+        $p = New-ConsoleApplication
+
+        $projectDirectoryPath = $p.Properties.Item("FullPath").Value
+        $packagesConfigPath = Join-Path $projectDirectoryPath 'packages.config'
+        $projectDirectoryPath = $p.Properties.Item("FullPath").Value
+        $solutionDirectory = Split-Path -Path $projectDirectoryPath -Parent   
+        # Write a file to disk, but do not add it to project
+        '<packages>
+            <package id="Contoso.MVC.ASP" version="1.0.0" targetFramework="net461" />
+    </packages>' | out-file $packagesConfigPath
+
+        CreateCustomTestPackage "Contoso.MVC.ASP" "1.0.0" $privateRepo "Thisisfromprivaterepo.txt"
+        CreateCustomTestPackage "Contoso.MVC.ASP" "1.0.0" $opensourceRepo "Thisisfromopensourcerepo.txt"
+
+        # Act
+        Build-Solution
+
+        # Assert   
+        $packagesFolder = Join-Path $solutionDirectory "packages"
+        $contosoNupkgFolder = Join-Path $packagesFolder "Contoso.MVC.ASP.1.0.0"
+        Assert-PathExists(Join-Path $contosoNupkgFolder "Contoso.MVC.ASP.1.0.0.nupkg")
+        # Make sure name squatting package from public repo not restored.
+        $contentFolder = Join-Path $contosoNupkgFolder "content"
+        Assert-PathExists(Join-Path $contentFolder "Thisisfromprivaterepo.txt")
```

Comment:
may be the best way to assert the package was truly downloaded from the expected source is to read content from `.nupkg.metadata` file? https://docs.microsoft.com/en-us/nuget/concepts/troubleshooting-installed-packages. Maybe I am missing some context here, please let me know.

Reply:
You're not missing anything, you're correct. Same thing can be done many way, I just did this way. I'll refactor this one to include `.nupkg.metadata` verification for when I test `install` action scenarios.

Reply:
Actually `source` in `.nupkg.metadata` is null for local sources, so it can't be used for asserting sources.

![image](https://user-images.githubusercontent.com/8766776/123502583-6bb16a80-d602-11eb-8af0-fe57735b9cc9.png)

Reply:
Actually `.nupkg.metadata` is not created for packages.config type solution's `packages` folder. 

![image](https://user-images.githubusercontent.com/8766776/123502939-ae744200-d604-11eb-807e-693e0b08101a.png)